tests/ui/codegen/equal-pointers-unequal/README.md

```diff
@@ -0,0 +1,22 @@
+See https://github.com/rust-lang/rust/issues/107975
+
+Basically, if you have two pointers with the same address but from two different allocations,
+the compiler gets confused whether their addresses are equal or not,
+resulting in some self-contradictory behavior of the compiled code.
+
+This folder contains some examples.
+They all boil down to allocating a variable on the stack, taking its address,
+getting rid of the variable, and then doing it all again.
+This way we end up with two addresses stored in two `usize`s (`a` and `b`).
+The addresses are (probably) equal but (definitely) come from two different allocations.
+Logically, we would expect that exactly one of the following options holds true:
+1. `a == b`
+2. `a != b`
+Sadly, the compiler does not always agree.
+
+Due to Rust having at least three meaningfully different ways
+to get a variable's address as an `usize`,
+each example is provided in three versions, each in the corresponding subfolder:
+1. `./as-cast/` for `&v as *const _ as usize`,
+2. `./strict-provenance/` for `addr_of!(v).addr()`,
+2. `./exposed-provenance/` for `addr_of!(v).expose_provenance()`.
```

Comment:
Thanks for the explanatory comments :) 

I'm not sure if creating a separate markdown file in the sub folder will cause some side effects (such as affecting the compiler-test tool). I would probably merge some of the cases(like `as-cast` cases) into one `.rs` file and put this comment at the top.

Reply:
> I'm not sure if creating a separate markdown file in the sub folder will cause some side effects (such as affecting the compiler-test tool).

I have not encountered any problems. Not to mention, this is not the first .md file there:

```
tests/ui/codegen/equal-pointers-unequal/README.md (the new one)
tests/ui/functions-closures/closure-expected-type/README.md
tests/ui/mir-dataflow/README.md
tests/ui/reachable/README.md
tests/ui/README.md
tests/ui/rustc-env/README.md
tests/ui/rustdoc/README.md
tests/ui/self/elision/README.md
tests/ui/specialization/README.md
tests/ui/specialization/README-rpass.md
```

> I would probably merge some of the cases(like `as-cast` cases) into one `.rs` file and put this comment at the top.

While composing these tests, I have observed that it is a bit finicky to trick the compiler into exhibiting this strange behavior, so I am a bit reluctant to merge these tests together, because I fear it might summon some nasal demons of inlining and codegen units, causing a different outcome.

Reply:
That's ok